tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java

```diff
@@ -55,8 +69,105 @@ public TLCSourceBreakpoint(final String module, final SourceBreakpoint s) {
+			suffix++;
+		} while (semanticRoot.getOpDef(bpOpName) != null);
+
+		// Then, send expression through syntax parser; wrap expression
+		// in module to make syntax parser happy, then throw away the
+		// encapsulating module. Future syntax parser updates might allow
+		// parsing the operator definition directly.
+		final String wrapper = "---- MODULE Ignore ----\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpOpName, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
+		TLAplusParser parser = new TLAplusParser(sourceCode, StandardCharsets.UTF_8.name());
+		boolean syntaxParseSuccess = parser.parse();
+		SyntaxTreeNode syntaxRoot = parser.ParseTree;
+		if (!syntaxParseSuccess || null == syntaxRoot) {
+			// Parse error is output to ToolIO.out
+			return null;
+		}
+
+		// Extract the operator definition syntax tree node. The unit
+		// definitions are always second element of the root children.
+		// We can reasonably infer that at least one unit-level op will
+		// exist given the wrapper text and parsing success.
+		SyntaxTreeNode syntaxOp = syntaxRoot.getHeirs()[2].getHeirs()[0];
+
+		// Run semantic analysis on the syntax tree node; this resolves
+		// its identifiers to those present in the spec, and also adds it
+		// to the spec's symbol table. Future work might make this method
+		// free of side-effects so as to skip this latter part. Since the
+		// spec's symbol table was modified we also need to invalidate
+		// the definition cache.
+		// There is a possibility (untested) that if a user performs a
+		// "get all references" type operation on some identifier in the
+		// spec, and this new breakpoint operator refers to that
+		// identifier, then the breakpoint operator could be returned in
+		// the list to the user and cause some kind of crash in the UI
+		// since it does not have an actual location in the spec.
+		OpDefNode semanticOp = null;
+		try {
+			semanticOp = semanticRoot.semanticChecker.processOperator(
+				syntaxOp,
+				null,
+				semanticRoot
+			);
+		} catch (AbortException e) {
+			ToolIO.err.print(e.toString());
+			ToolIO.err.print(semanticRoot.semanticChecker.errors.toString());
+			return null;
+		}
+		if (null == semanticOp || semanticRoot.semanticChecker.errors.isFailure()) {
+			ToolIO.err.print(semanticRoot.semanticChecker.errors.toString());
+			return null;
+		}
+		semanticRoot.clearOpDefCache();
+		
+		// Run level-checking. The operator should be restricted to
+		// action-level or below.
+		Errors levelCheckingErrors = new Errors();
+		SemanticNode.setError(levelCheckingErrors); // Annoyingly static
+		if (!semanticOp.levelCheck() || levelCheckingErrors.isFailure() || !semanticOp.levelCorrect) {
+			ToolIO.err.println(levelCheckingErrors.toString());
+			return null;
+		}
+		if (LevelConstants.ConstantLevel == semanticOp.level
```

Comment:
`tla2sany.semantic.LevelNode.getLevel()`

More importantly, can you make the context of the breakpoint determine the maximum level of the debug expression? For instance, the maximum level of a breakpoint expression for a constant formula is constant, and so on.

Related:
```java
tla2sany.semantic.SemanticNode.pathTo(Location)
```